File: Week 2/id_160/144.二叉树的前序遍历.cpp
/*
 * @lc app=leetcode.cn id=144 lang=cpp
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// class Solution {
// public:
//     vector<int> preorderTraversal(TreeNode* root) {
//         vector<int> v;
//         preTraversal(root, v);
//         return v;
// }
//     void preTraversal(TreeNode* root, vector<int>& v){
//         if(!root) return;
//         v.push_back(root->val);
//         preTraversal(root->left, v);
//         preTraversal(root->right, v);
//     }
// };

// class Solution {
// public:
//     vector<int> preorderTraversal(TreeNode* root) {
//         vector<int> v;
//         if(!root) return v;
//         TreeNode* temp = root;
//         stack<TreeNode*> s;
//         s.push(root);
//         while(!s.empty()){
//             temp = s.top();
//             s.pop();
//             v.push_back(temp->val);
//             if(temp->right) s.push(temp->right);
//             if(temp->left) s.push(temp->left);
//         }
//         return v;
//     }
// };

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> v;
        if(!root) return v;
        TreeNode* temp = root, *prev;
        while(temp){
            if(!temp->left){
                v.push_back(temp->val);
                temp = temp->right;
            }else{
                prev = temp->left;
                while(prev->right&&(prev->right != temp))
                    prev = prev->right;
                if(!prev->right){
                    v.push_back(temp->val);
                    prev->right = temp;
                    temp = temp->left;
                }else{
                    prev->right = NULL;
                    temp = temp->right;
                }
            }
        }
        return v;
    }
};
// @lc code=end

File: Week 2/id_520/144.二叉树的前序遍历.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=144 lang=java
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
    //     //方法一，递归,但是题目建议使用迭代
    //     List<Integer> preResult = new ArrayList<>();
    //     preHelper(preResult, root);
    //     return preResult;
    // }
    // public void preHelper(List<Integer> preResult, TreeNode root) {
    //     if (root == null) return;
    //     preResult.add(root.val);
    //     preHelper(preResult, root.left);
    //     preHelper(preResult, root.right);
    // }

        // //方法二，迭代, 来自lLeetCode高票代码
        // List<Integer> result = new ArrayList<>();
        // Stack<TreeNode> stack = new Stack<>();
        // while(root != null || !stack.empty()) {
        //     result.add(root.val);
        //     if (root.right != null) stack.push(root.right);
        //     root = root.left;
        //     if (root == null && !stack.isEmpty()) root = stack.pop();
        // }
        // return result;

        //方法三，非递归的LeetCode最高票代码，前中后序代码相对来说最好记，因为有规律
        LinkedList<Integer> result = new LinkedList<>();
        Stack<TreeNode> stack = new Stack<>();
        while (!stack.empty() || root != null) {
            if (root != null) {
                stack.push(root);
                result.add(root.val);   //Add before going to children
                root = root.left;
            } else {
                TreeNode node = stack.pop();
                root = node.right;
            }
        }
        return result;
    }
}
// @lc code=end

